pkg/operator2/operator.go

```diff
@@ -217,6 +225,11 @@ func (c *authOperator) Sync(obj metav1.Object) error {
 
 		// manually update the conditions while preserving last transition time
 		for _, condition := range operatorConfigCopy.Status.Conditions {
+			// Avoid setting conditions that are managed by other
+			// controllers to avoid racing with them.
+			if c.ignoredConditionTypes.Has(condition.Type) {
+				continue
+			}
 			v1helpers.SetOperatorCondition(&status.Conditions, condition)
```

Comment:
This would just delete those conditions that you keep ignored instead, wouldn't it?

Besides, I like to think of controllers as of separate logical units. I think the right thing to do would be to have `handleSync` return a slice of conditions/conditionUpdateFunctions that you could just pass to `UpdateStatus`. It would just require a minor change to `handleDegraded` and passing those updates up. WDYT?

Reply:
> This would just delete those conditions that you keep ignored instead, wouldn't it?

I don't believe that is the case. `status` is retrieved from the informer, and includes all conditions set by all controllers. The code here sets/updates only those conditions managed by this controller, and does not delete/update conditions set by other controllers.

> Besides, I like to think of controllers as of separate logical units. I think the right thing to do would be to have `handleSync` return a slice of conditions/conditionUpdateFunctions that you could just pass to `UpdateStatus`. It would just require a minor change to `handleDegraded` and passing those updates up. WDYT?

What you are proposing is the strategy already used by the ingress state and router certs controllers. I investigated implementing it for the primary controller, but it looked far more invasive due to the need to plumb collection and propagation of update functions through the entire callchain. Given that a refactor of the operator is ongoing, does it make sense to introduce such a change as part of that refactor rather than in the fixing of this bug?

Reply:
> The code here sets/updates only those conditions managed by this controller, and does not delete/update conditions set by other controllers.

don't know what I was thinking...

> Given that a refactor of the operator is ongoing, does it make sense to introduce such a change as part of that refactor rather than in the fixing of this bug?

this is a rare chance we get to do the right thing now rather than to write it down and postpone it to somewhen, I think it's a good excuse to do it.

Reply:
It's going to be really invasive. Happy to do it as a follow-up, but I really think it's orthogonal to the fix proposed here and much riskier.

Reply:
Also - the fix proposed here is trivial to backport (and it should be backported to 4.4 and 4.3). I'm less convinced of the wisdom of backporting the fix you are proposing.

Reply:
Why would we have to backport it to 4.3 and 4.2 when ingress state controller was added in 4.4? It sure would not make sense to backport it all the way.

Also, the idea was to just make `handleDegraded` return a condition which could then bubble all the way up to `Sync`. I am apparently missing something?